api/USERS_GUIDE.md

```diff
@@ -281,4 +281,59 @@ current holds for the project:
     "P1.proj_waiting",
     "P1.proj_avail"
 ]
-```
\ No newline at end of file
+```
+
+### Proofreading
+
+The proofreading APIs require retaining and passing in state information. The
+following examples use [jq](https://jqlang.github.io/jq/) to parse responses
+and store them as environment variables.
+
+Get a project's details and parse out the project state:
+```bash
+curl -s -H "Accept: application/json" -H "X-API-KEY: $API_KEY" \
+    -X PUT "https://www.pgdp.org/api/v1/projects/projectID44de3936807f1" \
+    > project.json
+
+PROJECT_STATE=$(cat project.json | jq -r .state)
+```
+
+Check out a page and parse out the page name, state, and page text:
+```bash
+curl -s -H "Accept: application/json" -H "X-API-KEY: $API_KEY" \
+    -X PUT "https://www.pgdp.org/api/v1/projects/projectID44de3936807f1/checkout?state=$PROJECT_STATE" \
+    > proof_page.json
+
+PAGE_NAME=$(cat proof_page.json | jq -r .pagename)
+PAGE_STATE=$(cat proof_page.json | jq -r .pagestate)
+cat proof_page.json | jq '{text}' > page_text.json
+```
+
+Validate the page text doesn't have any weird characters:
```

Comment:
This is fine but I'd intended this to be used only if you had found there were weird characters when save (as in progress or as done) returned an error saying there were weird characters. It returns not only vlaid as true or false but the text marked up to show the weird characters.

Reply:
For my own information, if the only extra piece of information that is returned by the separate `validate` function is the list of weird characters, why not do that as part of the save process?

Reply:
There are really two questions here:
1. Why not return the weird character information when saving?
2. Why not just return a list of weird characters instead of a marked up text?

For the first question, you could equally well do it like that and some earlier versions did and it would be more efficient in the sense that you do not have to submit the text twice. The only reason not to is that if there are weird characters the function will fail to save the submitted text so it should arguably return a 400+ code. You can still return the information with a 400+ code but it seemed simpler not to. There are no good reasons for not doing it like that if you wanted to.

The second question I had not considered as such. The way the present system works is that the client gets a regular expression from the server to use to check for invalid characters which seems a bit messy because javascrpt regexs work a bit differently to php regexs so it seemed better to let the server do the work and send the result back to the client. If you want to display a text with the characters highlighted then if you just return a list of characters then the client would have to search for them which is possible but just a different way to do it. There is also the question of complicated situations involving multiple unicode codepoints and graphemes where the client and server would have to agree on exactly what constituted a grapheme. If it was a large text then it could be better to do it like that so as not to have to return so much data.

Reply:
Yeah, predictability of regex engines can trip us up if we're not careful. Thanks for the explanation. I suspect we could get into an interesting philosophical discussion about the whole issue, but I don't think this is the place to. 😁